clients/pkg/promtail/targets/heroku/target.go

```diff
@@ -0,0 +1,177 @@
+package heroku
+
+import (
+	"flag"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	herokuEncoding "github.com/heroku/x/logplex/encoding"
+	"github.com/imdario/mergo"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/model/relabel"
+	"github.com/weaveworks/common/server"
+
+	"github.com/grafana/loki/clients/pkg/promtail/api"
+	"github.com/grafana/loki/clients/pkg/promtail/scrapeconfig"
+	"github.com/grafana/loki/clients/pkg/promtail/targets/target"
+
+	"github.com/grafana/loki/pkg/logproto"
+	util_log "github.com/grafana/loki/pkg/util/log"
+)
+
+type Target struct {
+	logger         log.Logger
+	handler        api.EntryHandler
+	config         *scrapeconfig.HerokuTargetConfig
+	jobName        string
+	server         *server.Server
+	metrics        *Metrics
+	relabelConfigs []*relabel.Config
+}
+
+func NewTarget(metrics *Metrics, logger log.Logger, handler api.EntryHandler, jobName string, config *scrapeconfig.HerokuTargetConfig, relabel []*relabel.Config) (*Target, error) {
+
+	ht := &Target{
+		metrics:        metrics,
+		logger:         logger,
+		handler:        handler,
+		jobName:        jobName,
+		config:         config,
+		relabelConfigs: relabel,
+	}
+
+	// Bit of a chicken and egg problem trying to register the defaults and apply overrides from the loaded config.
+	// First create an empty config and set defaults.
+	defaults := server.Config{}
+	defaults.RegisterFlags(flag.NewFlagSet("empty", flag.ContinueOnError))
+	// Then apply any config values loaded as overrides to the defaults.
+	if err := mergo.Merge(&defaults, config.Server, mergo.WithOverride); err != nil {
+		level.Error(logger).Log("msg", "failed to parse configs and override defaults when configuring push server", "err", err)
+	}
+	// The merge won't overwrite with a zero value but in the case of ports 0 value
+	// indicates the desire for a random port so reset these to zero if the incoming config val is 0
+	if config.Server.HTTPListenPort == 0 {
+		defaults.HTTPListenPort = 0
+	}
+	if config.Server.GRPCListenPort == 0 {
+		defaults.GRPCListenPort = 0
+	}
+	// Set the config to the new combined config.
+	config.Server = defaults
+
+	err := ht.run()
+	if err != nil {
+		return nil, err
+	}
+
+	return ht, nil
+}
+
+func (h *Target) run() error {
+	level.Info(h.logger).Log("msg", "starting heroku target", "job", h.jobName)
+	// To prevent metric collisions because all metrics are going to be registered in the global Prometheus registry.
+	h.config.Server.MetricsNamespace = "promtail_heroku_target_" + h.jobName
+
+	// We don't want the /debug and /metrics endpoints running
+	h.config.Server.RegisterInstrumentation = false
```

Comment:
WeaveWorks http server (the one used in pushtarget an here) has a feature to expose debug and metrics capabilities. Since this is not a "central" promtail http server but one just used for a specific ingress (heroku logs) we just one to expose the least functionality and surface we can, hence disabling this extra-features.